Add trafikverket_weatherstation sensor platform

Qxlkdr opened the pull request:
Created PR 12111 but due to permission issue I'm creating a new fork and PR.

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4562

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: trafikverket_weatherstations
    name: Trafikverket Road WeatherStation Kungälv
    api_key: eXXcbXXXacXXXXc39XX3aXXX4aXX46XX
    station: Kungälv
    type: road
```

## Checklist:
  - [X] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Qxlkdr committed:
Create trafikverket_weatherstation.py

Created PR 12111 but due to permission issue I'm creating a new fork and PR.

Qxlkdr commented:
As I can see there was two failed Travis tests.

**First**
```
C:120, 8: Invalid variable name "lj" (invalid-name)
C:123, 8: Invalid variable name "mm" (invalid-name)
```

**Second**
```
./homeassistant/components/sensor/trafikverket_weatherstation.py:1:1: D400 First line should end with a period
```

Both are related to Python 3.4.2. Do I need to correct them somehow or will we pass them? I don't see how the D400 issue should be corrected. All Python files begin with:

```
"""
Weather information for air and road temperature, provided by Trafikverket

For more details about this platform, please refer to the documentation at
https://home-assistant.io/components/sensor.trafikverket_weatherstation/
"""
```

Qxlkdr committed:
Added dot

Added dot to the first (second) row of the file, after the description.

Danielhiversen commented:
They are invalid variable names.
You should give them a longer and more descriptive name

Danielhiversen commented:
You also have to add your file to .coveragerc as described in the pull request template

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+
+CONF_API = 'api'
+CONF_STATION = 'station'
+CONF_TYPE = 'type'
```
from homeassistant.const import CONF_TYPE

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+
+CONF_ATTRIBUTION = "Data provided by Trafikverket API"
+
+CONF_API = 'api'
```
from homeassistant.const import CONF_API_KEY

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+        This is the only method that should fetch new data for Home Assistant.
+        """
+        import requests
+        import json
```
These two imports are standard libraries and should be imported at the top of the file

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+                  </QUERY>
+            </REQUEST>"""
+
+        post = requests.post(url, data=xml.encode('utf-8'))
```
What happens if you do not have an internet connection or the request fails?

fabaff replied:
All requests should have a timeout.

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+        lj = json.loads(post.text)
+
+        # mm = measurement
+        mm = lj["RESPONSE"]["RESULT"][0]["WeatherStation"][0]["Measurement"]
```
Make this more fail safe. What if WeatherStation does not exists in the response?

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_API): cv.string,
+    vol.Required(CONF_STATION): cv.string,
```
Do you have a list of valid stations?

Qxlkdr replied:
Yes, it's written in the documentation under header "See Trafikverket weather stations":
https://github.com/home-assistant/home-assistant.github.io/pull/4562/files

In the map you will need to click a choosen station and then copy it's name, like highlighted:
https://imgur.com/a/TGQKC

Thanks for your review and notice of the other issues. I will try to fix them tomorrow or during the weekend. Br Qxlkdr

Danielhiversen replied:
So what happens if the user provides an invalid station or api token?

Qxlkdr replied:
I will try to fix that. However, do I need to have an error handler for both cases? Or can I just check for a int/double? If the result variable doesn't contains an int/double the has been an error in the request. As I see it, it could only be 3 cases:
1. No network
2. Invalid station
3. Invalid API key

I can have an error handler for number 2-3 (checking for int/double). If not received an error message like "Cannot fetch station, check station name and API key" in the log. Number 1 can be checked separately. So, 3 issues, 2 solutions to all 3 of them. Is that ok or will I need to have an error handler for all 3, separately?

Danielhiversen replied:
The most important is that the code does not crash in any of these cases.
And it would be great if you can provide a log message to help the user fix the problem.

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+                        <FILTER>
+                              <EQ name="Name" value='""" + self._station + """' />
+                        </FILTER>
+                        <INCLUDE>Measurement.Road.Temp</INCLUDE>
```
Have you tried something like this? 
`<INCLUDE>Measurement.'""" + 'Air'  if self._type == 'air' else 'Road' + """'.Temp</INCLUDE>`

Qxlkdr committed:
Corrections based on feedback

Done:
- Run flake8 before this commit
- Fixed invalid variables
- Shortened the xml variable/query via if statement (air_vs_road)
- Moved imports if update() to top of the file
- Imported CONF_API_KEY and CONF_TYPE
- Updated documentation (api_key): home-assistant/home-assistant.github.io#4562

Actions left:
- Error handling
- Request timeout
- Add sensor (file) to .coveragerc

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,120 @@
+        # mea = measurement
+        mea = loa["RESPONSE"]["RESULT"][0]["WeatherStation"][0]["Measurement"]
+        if self._type == 'air':
+            result = mea["Air"]["Temp"]
```
Then you could use air_vs_road here also

Qxlkdr replied:
I'm sorry, where could I use `air_vs_road` here and why? You lost me here.

Danielhiversen replied:
Instead of the if sentence:
`result = mea[air_vs_road]["Temp"]`

Danielhiversen replied:
or 
`self._state = loa["RESPONSE"]["RESULT"][0]["WeatherStation"][0]["Measurement"][air_vs_road]["Temp"]`

Qxlkdr replied:
Sorry, I was totally lost, ofc. Thanks for the suggestion. Saves unnecessary memory.

Danielhiversen commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,120 @@
+              </QUERY>
+        </REQUEST>"""
+
+        post = requests.post(url, data=xml.encode('utf-8'))
```
Add a timeout

Qxlkdr committed:
Multiple corrections

Done:
- Executed pylint and flake8 tests before commit
- Fixed import order
- Implemented request timeout
- Used variable air_vs_road in the return as well

Actions left:
- Error handling
- Add sensor (file) to .coveragerc

Qxlkdr committed:
Error handling

Done:
- Error handling for network
- Error handling for JSON response

Qxlkdr committed:
Added trafikverket_weatherstation.py

Added trafikverket_weatherstation.py in correct order.

Qxlkdr committed:
Road as default

Changed if statement to check for 'road' which means it will always defaulting to 'air' in other cases. Even if it will only accept 'air' and 'road' from the PLATFORM_SCHEMA.

Qxlkdr commented:
"All checks have passed" and I think I have fixed all the "requested changes".

Danielhiversen reviewed (approved):
:)

Qxlkdr commented:
@Danielhiversen What's next? Do I need to do anything more? What about the documentation?

Danielhiversen commented:
Nothing more for you to do at the moment. :)
Someone else will take a look and merge it if they are happy with the code.

Qxlkdr commented:
@Danielhiversen Perfect! Thanks for the help and good reviewing job. :)

Qxlkdr commented:
@Danielhiversen @fabaff 
Will this component be added/merged to the coming release? Sorry for the inconvenience, this is my first component, PR and interaction with the team. The documentation is updated as well. Thanks in advance. Have a nice weekend.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+            post = requests.post(url, data=xml.encode('utf-8'), timeout=5)
+        except requests.exceptions.RequestException as err:
+            _LOGGER.error("Please check network connection: %s", err)
+            return None
```
Nothing is checking this return value.

Qxlkdr replied:
What do you mean? I don't even know if a return is needed here? The only thing I want is to write the error message in the log.

MartinHjelmare replied:
I think it's good to return here, since if the request fails no further data parsing is needed, and we shouldn't update state. But you should not `return None`. Just:
```py
return
```

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+        # Checking JSON respons.
+        try:
+            # loa (load) = loaded json
+            loa = json.loads(post.text)
```
Change the variable name to data, or similar.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+            loa = json.loads(post.text)
+
+            # mea = measurement
+            mea = loa["RESPONSE"]["RESULT"][0]
```
Please write out the variable names, so that the are easily understandable. You could use `result` here if you don't want to use `measurement`.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+            wea = mea["WeatherStation"][0]["Measurement"]
+        except KeyError:
+            _LOGGER.error("Incorrect weather station or API key.")
+            return None
```
See above.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+        """Return the state attributes."""
+        return self._attributes
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
```
This will only throttle this entity instance method. If a user sets up multiple platforms for different stations, those will each hit the api endpoint, ie the min time between updates will be less than what this throttle allows.

So this throttle is only effective as a protection from the user setting a custom scan interval that is too low, in the one platform scenario.

All of this might not be a problem, but it's good to think about.

Qxlkdr replied:
This was the only reason. Protection against custom scan intervals that is too low.

Qxlkdr committed:
Updated variable names

Updated variable names to be more understandable as requested by @MartinHjelmare

Qxlkdr commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,129 @@
+import json
+import logging
+from datetime import timedelta
+import requests
```
If you mean requests isn't standard, then I have added an extra blank line. Or did you mean anything else?

Qxlkdr committed:
Standard Libraries

Grouped Standard Libraries

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -0,0 +1,124 @@
+            final = result["WeatherStation"][0]["Measurement"]
+        except KeyError:
+            _LOGGER.error("Incorrect weather station or API key.")
+            return None
```
Don't return None, just:
```py
return
```

Qxlkdr replied:
Done, what do you think? All changes approved? Thanks in advance.

Qxlkdr committed:
Return None

Changed return None to only return as suggested.

MartinHjelmare reviewed (approved):
Looks good!

Qxlkdr commented:
@Danielhiversen @fabaff @MartinHjelmare 
What is the criteria for getting a PR merged? This one has been in status "Ready for review" since 9 Feb, almost 6 weeks now. Are we waiting for the documentation to get approved or have the team forgot it? 😄

MartinHjelmare commented:
Sorry about the delay.

Qxlkdr commented:
Thanks a lot @MartinHjelmare 😃
